### PR TITLE
Avoid sqrt calculations with values ​​less than zero

### LINKED ISSUE #135542: torch.nn.InstanceNorm2d returns nan with negative eps values for tensor with float16 dtype on cpu
### 🐛 Describe the bug

When I run this code on a tensor of type float16 of shape (1, 3, 256, 256) on torch.nn.InstanceNorm2d, I get a tensor with nan values from cpu but gpu can produce non-nan tensor. I am providing the tensor in a pickle file which can be downloaded from here: [link](https://drive.google.com/file/d/1-Qm6x7APyxGWSu92svOY3l_mBirlgsdp/view?usp=drive_link)

Minimal repro:

```python
import torch
import pickle

with open('input_instancenorm2d.pickle', 'rb') as f:
  input_dict = pickle.load(f)

a = torch.tensor(input_dict['input'], dtype=torch.float16)
print(a.shape)
# torch.Size([1, 3, 256, 256])
print(a.dtype)
# torch.float16
num_features = 3
eps = -0.5
m = torch.nn.InstanceNorm2d(num_features=num_features, eps=eps)
output_cpu = m(a)
output_gpu = m.cuda()(a.cuda())
print(output_cpu[0][0][0][0])
# tensor(nan, dtype=torch.float16)
print(output_gpu[0][0][0][0])
# tensor(-1.0039, device='cuda:0', dtype=torch.float16)
```

It can also be reproduced on [colab](https://colab.research.google.com/drive/1X0AowednmpH79htWyiti3V8gG80pvpwq?usp=sharing). Just need to upload the pickle file to the files tab after connecting to a runtime.

Note: I found the tensor with fuzzing.

### Versions

Collecting environment information...
PyTorch version: 2.4.0+cu121
Is debug build: False
CUDA used to build PyTorch: 12.1
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.6 LTS (x86_64)
GCC version: (Ubuntu 9.4.0-1ubuntu1~20.04.2) 9.4.0
Clang version: 15.0.0 (git@github.com:llvm/llvm-project.git 4ba6a9c9f65bbc8bd06e3652cb20fd4dfc846137)
CMake version: version 3.22.1
Libc version: glibc-2.31

Python version: 3.10.0 (default, Mar  3 2022, 09:58:08) [GCC 7.5.0] (64-bit runtime)
Python platform: Linux-5.15.0-117-generic-x86_64-with-glibc2.31
Is CUDA available: True
CUDA runtime version: Could not collect
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: GPU 0: NVIDIA GeForce RTX 4090
Nvidia driver version: 555.42.02
cuDNN version: /usr/lib/x86_64-linux-gnu/libcudnn.so.7.6.5
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                         x86_64
CPU op-mode(s):                       32-bit, 64-bit
Byte Order:                           Little Endian
Address sizes:                        39 bits physical, 48 bits virtual
CPU(s):                               24
On-line CPU(s) list:                  0-23
Thread(s) per core:                   1
Core(s) per socket:                   16
Socket(s):                            1
NUMA node(s):                         1
Vendor ID:                            GenuineIntel
CPU family:                           6
Model:                                183
Model name:                           13th Gen Intel(R) Core(TM) i7-13700KF
Stepping:                             1
CPU MHz:                              5333.317
CPU max MHz:                          5800.0000
CPU min MHz:                          800.0000
BogoMIPS:                             6835.20
Virtualization:                       VT-x
L1d cache:                            384 KiB
L1i cache:                            256 KiB
L2 cache:                             16 MiB
NUMA node0 CPU(s):                    0-23
Vulnerability Gather data sampling:   Not affected
Vulnerability Itlb multihit:          Not affected
Vulnerability L1tf:                   Not affected
Vulnerability Mds:                    Not affected
Vulnerability Meltdown:               Not affected
Vulnerability Mmio stale data:        Not affected
Vulnerability Reg file data sampling: Mitigation; Clear Register File
Vulnerability Retbleed:               Not affected
Vulnerability Spec rstack overflow:   Not affected
Vulnerability Spec store bypass:      Mitigation; Speculative Store Bypass disabled via prctl and seccomp
Vulnerability Spectre v1:             Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:             Mitigation; Enhanced / Automatic IBRS; IBPB conditional; RSB filling; PBRSB-eIBRS SW sequence; BHI BHI_DIS_S
Vulnerability Srbds:                  Not affected
Vulnerability Tsx async abort:        Not affected
Flags:                                fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf tsc_known_freq pni pclmulqdq dtes64 monitor ds_cpl vmx est tm2 ssse3 sdbg fma cx16 xtpr pdcm sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb ssbd ibrs ibpb stibp ibrs_enhanced tpr_shadow vnmi flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid rdseed adx smap clflushopt clwb intel_pt sha_ni xsaveopt xsavec xgetbv1 xsaves split_lock_detect avx_vnni dtherm ida arat pln pts hwp hwp_notify hwp_act_window hwp_epp hwp_pkg_req umip pku ospke waitpkg gfni vaes vpclmulqdq rdpid movdiri movdir64b fsrm md_clear serialize arch_lbr flush_l1d arch_capabilities

Versions of relevant libraries:
[pip3] numpy==1.26.4
[pip3] optree==0.11.0
[pip3] torch==2.4.0
[pip3] triton==3.0.0
[conda] numpy                     1.26.4          py310heeff2f4_0  
[conda] numpy-base                1.26.4          py310h8a23956_0  
[conda] optree                    0.11.0                   pypi_0    pypi
[conda] torch                     2.4.0                    pypi_0    pypi
[conda] triton                    3.0.0                    pypi_0    pypi

### PR DESCRIPTION
**Pull request by CaoE**
Fixes #135542 and #136680.

### CODE REVIEW

**Review by jgong5 — APPROVED** [review]
Add UT?

File: aten/src/ATen/native/Normalization.cpp [hunk 1]
```diff
@@ -103,11 +103,7 @@ namespace {
 template<typename T>
 struct InvStd {
   T operator()(T var, double epsilon) const {
-    T invstd = 0;
-    if (var != static_cast<T>(0) || epsilon != static_cast<T>(0)) {
-      invstd = static_cast<T>(1) / std::sqrt(var + epsilon);
-    }
-    return invstd;
+    return static_cast<T>(1) / std::sqrt(var + epsilon);
```
Comment (jgong5) [inline aten/src/ATen/native/Normalization.cpp:106]: hmm, why do you change things like this?

**Review by jgong5 — CHANGES_REQUESTED** [review]

File: aten/src/ATen/native/Normalization.cpp (hunk 1 above)
Comment (CaoE) [reply aten/src/ATen/native/Normalization.cpp:106]: CUDA directly calculates `c10::cuda::compat::rsqrt(var + eps)`.
I do this change to align it with cuda.
If var + eps <= 0 is a meaningless input and we check `var + eps <= 0`,  users might not be able to aware that the input is meaningless.

Comment (CaoE) [reply aten/src/ATen/native/Normalization.cpp]: Change it back.

**Review by jgong5 — APPROVED** [review]

**Comment by CaoE** [comment]
@Skylion007 Could you please review this PR ? Thanks.

**Comment by CaoE** [comment]
@Skylion007 Could you please review this PR ? Thanks.

**Comment by CaoE** [comment]
@Skylion007 Could you please review this PR ? Thanks.

**Comment by CaoE** [comment]
@cpuhrsch Could you please review this PR ? Thanks.

**Review by cpuhrsch — APPROVED** [review]

**Review by cpuhrsch — CHANGES_REQUESTED** [review]
Sorry, let's wait a bit longer. Do we even want to allow negative eps in general? If not, maybe this check needs to be done higher up. I'd also expect this addition could technically change the performance (but likely doesn't matter). However, since this is a very core operator I'd want to be extra sure. Disallowing negative eps would fix this too afaict. cc @mikaylagawarecki @albanD

**Comment by albanD** [comment]
> Do we even want to allow negative eps in general? 

I would definitely agree with forbidding that. As the issue mentions, this behavior was only discovered by fuzzing, not a real use case.

That being said, checking equality with 0 is definitely a footgun as well, so the change in this PR does make sense, I would even go one step further and use `max(var, epsilon)` to avoid biasing the user value.

**Comment by CaoE** [comment]
@albanD @cpuhrsch CUDA directly calculates `c10::cuda::compat::rsqrt(var + eps)` without checks. Do we need to align with cuda? Or do we need to throw messages to user if eps or val <= 0. 
BTW, I have a question about how we define what kind of input is meaningful, for example: var > 0 or eps > 0 or var + eps > 0 ?
If var + eps <= 0 or val <=0 or eps  <=0  are meaningless inputs and we return a value of 0 directly, users might not be able to aware that the input is meaningless.

**Comment by CaoE** [comment]
> @albanD @cpuhrsch CUDA directly calculates `c10::cuda::compat::rsqrt(var + eps)` without checks. Do we need to align with cuda? Or do we need to throw messages to user if eps or val <= 0. BTW, I have a question about how we define what kind of input is meaningful, for example: var > 0 or eps > 0 or var + eps > 0 ? If var + eps <= 0 or val <=0 or eps <=0 are meaningless inputs and we return a value of 0 directly, users might not be able to aware that the input is meaningless.

Hi @albanD @cpuhrsch Can we follow CUDA directly currently and decide whether we need disallow negative eps or throw msg to user later ?
